src/cose/test_cose.rs

```diff
@@ -22,43 +23,115 @@ const P256_PARAMS: SignatureParameters = SignatureParameters {
     algorithm: SignatureAlgorithm::ES256,
     pkcs8: &test::PKCS8_P256_EE,
 };
+const P384_PARAMS: SignatureParameters = SignatureParameters {
+    certificate: &test::P384_EE,
+    algorithm: SignatureAlgorithm::ES384,
+    pkcs8: &test::PKCS8_P384_EE,
+};
+const P521_PARAMS: SignatureParameters = SignatureParameters {
+    certificate: &test::P521_EE,
+    algorithm: SignatureAlgorithm::ES521,
+    pkcs8: &test::PKCS8_P521_EE,
+};
 
-#[test]
-fn test_cose_sign_verify() {
+#[cfg(test)]
+fn test_verify(payload: &[u8], cert_chain: &[&[u8]], params_vec: Vec<SignatureParameters>) {
     test::setup();
-    let payload = b"This is the content.";
-    let certs: [&[u8]; 2] = [&test::P256_ROOT,
-                             &test::P256_INT];
-    let params_vec = vec![P256_PARAMS];
-    let cose_signature = sign(payload, &certs, &params_vec);
+    let cose_signature = sign(payload, &cert_chain, &params_vec);
     assert!(cose_signature.is_ok());
     let cose_signature = cose_signature.unwrap();
 
     // Verify signature.
     assert!(verify_signature(payload, cose_signature).is_ok());
 }
 
-#[test]
-fn test_cose_sign_verify_modified_payload() {
+#[cfg(test)]
+fn test_verify_modified_payload(
+    payload: &mut [u8],
+    cert_chain: &[&[u8]],
+    params_vec: Vec<SignatureParameters>,
+) {
     test::setup();
-    let payload = b"This is the content.";
-    let certs: [&[u8]; 2] = [&test::P256_ROOT,
-                             &test::P256_INT];
-    let params_vec = vec![P256_PARAMS];
-    let cose_signature = sign(payload, &certs, &params_vec);
+    let cose_signature = sign(payload, &cert_chain, &params_vec);
     assert!(cose_signature.is_ok());
     let cose_signature = cose_signature.unwrap();
 
     // Verify signature.
-    let payload = b"This is the content!";
+    payload[0] = !payload[0];
     let verify_result = verify_signature(payload, cose_signature);
     assert!(verify_result.is_err());
     assert_eq!(verify_result, Err(CoseError::VerificationFailed));
 }
 
+#[cfg(test)]
+fn test_verify_modified_signature(
+    payload: &[u8],
+    cert_chain: &[&[u8]],
+    params_vec: Vec<SignatureParameters>,
+) {
+    test::setup();
+    let cose_signature = sign(payload, &cert_chain, &params_vec);
+    assert!(cose_signature.is_ok());
+    let mut cose_signature = cose_signature.unwrap();
+
+    // Tamper with the cose signature.
+    cose_signature[15] = cose_signature[15] ^ cose_signature[15];
+
+    // Verify signature.
+    let verify_result = verify_signature(payload, cose_signature);
+    assert!(verify_result.is_err());
+    assert_eq!(verify_result, Err(CoseError::VerificationFailed));
+}
+
+#[cfg(test)]
+fn test_verify_verification_fails(
```

Comment:
Maybe comment that this is to be used in cases where the signature will fail to verify because the params aren't consistent? (i.e. it says it's signing with one certificate but it's actually using a private key that doesn't correspond to that cert)